USART_DMACmd test

ledvinap opened the pull request:
Can you test this code?
its USART_DMACmd + minor code cleanup ...

ezshinoda committed:
Pragmatic solution (from #1968)

Protect uartStartTxDMA from reentry

Move active transfer check into ATOMIC_BLOCK

Move all activity into (renamed) uartTryStartDMA

Reverted DMA_Cmd(DISABLE) for F4

wip

Reading NDTR to detect premature termination

Restart if NDTR != 0

Re-enable if CNDTR != 0 (F1, untested)

Removed bogus disable (F4 only)

Regression for F1 and F3

Rename uartStartTxDMA to uartTryStartTxDMA

serial_uart.h now defines uartTryStartTxDMA.
Untested, but should be okay (name change only).

For FEIF, just clear it and it should resume ...

ledvinap committed:
USART_DMACmd test

jflyper commented:
First commit: I still get lots of underruns, and configurator is complaining about various errors once again.

Trying the second one...

Ah, I already had this removed with the 1st commit. Same symptom.

ledvinap commented:
And the error is FEIF only ? or are other flags set too ?

jflyper commented:
Ah, I have to re-install the debug counter again to check other errors.

But there's something else that is very interesting.

I added the reenable code again, and guess what, I only see occasional FEIF then.

I will push the branch separately.

ledvinap commented:
Using pins for trace (+saleae) may be helpful ... it may reveal when the error occurs ..

(https://github.com/ledvinap/cleanflight/blob/local/src/main/drivers/pin_debug.h, but that ais a bit of overkill)

ledvinap commented:
'reenable code' ? What do you mean?

jflyper commented:
At the beginning of the uartTryStartTxDMA:

```
        if (s->txDMAStream->NDTR)
            goto reenable;
```
then jumps to USART_DMACmd(DISABLE).

jflyper commented:
[This is the working branch.](https://github.com/jflyper/cleanflight/tree/bfdev-fixing-txdma-with-ledvinap)

ledvinap commented:
Yes, enabling DMA again should mask the problem (probably even without corrupting data, FIFO error is raised befor UART register is written). But what is causing the buffer underflow?

ledvinap commented:
What happens when DMA priority is increased? https://github.com/jflyper/cleanflight/blob/0cc47aa1a6377adf3dd30888584ac0308120c851/src/main/drivers/serial_uart.c#L162

ledvinap commented:
(sorry, I have to leave today)

jflyper commented:
That's for rx... you meant tx, right?

jflyper commented:
Thanks for helping me!!! 👍  I'll get some sleep also. 😴

ledvinap commented:
Yes, Tx channel ...

ledvinap committed:
Change DMA_GET_FLAG_STATUS to return flags

Replace complicated macros with inline functions.
Shift returned flags so returning multiple flags is useful. This change
is backward-compatible

ledvinap committed:
refactor code a bit

ledvinap committed:
Use new DMA flags in usart handler

ledvinap commented:
Pushed few changes, mostly cosmetic.
Removed code that disables USART DMA before enabling DMA

ledvinap committed:
Fix DMA copy&paste error

ezshinoda committed:
For FEIF, just clear it and it should resume ...

ledvinap committed:
Revers some rebase artefacts

jflyper committed:
Merge branch 'bfdev-fix-uart-txdma-underrun' into bfdef-fix-uart-txdma-corruption
